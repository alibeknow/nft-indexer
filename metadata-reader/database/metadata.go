package database

import (
	"context"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"

	"metadata-reader/metadata"
)

type MetadataDB struct {
	collection *mongo.Collection
}

func (mDB MetadataDB) InsertOrUpdate(ctx context.Context, m metadata.Metadata) error {
	currentTime := time.Now()

	filter := bson.M{"_id": m.ID}
	update := bson.M{
		"$set": bson.M{
			"metadata":   m.Metadata,
			"type":       m.Type,
			"updated_at": currentTime,
		},
		"$setOnInsert": bson.M{
			"created_at": currentTime,
		},
	}

	opts := options.Update().SetUpsert(true)
	_, err := mDB.collection.UpdateOne(ctx, filter, update, opts)
	return err
}
