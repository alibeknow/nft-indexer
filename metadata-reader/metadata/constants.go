package metadata

type Chain string

const (
	ETH     Chain = "eth"
	POLYGON Chain = "polygon"
)

func GetMappings() map[string]interface{} {
	properties := make(map[string]interface{})

	properties["name"] = map[string]interface{}{
		"type":  "text",
		"index": true,
	}

	properties["description"] = map[string]interface{}{
		"type":  "text",
		"index": true,
	}

	properties["attributes"] = map[string]interface{}{
		"type":    "object",
		"enabled": true,
		"dynamic": true,
		"properties": map[string]interface{}{
			"key": map[string]interface{}{
				"type": "text",
			},
			"value": map[string]interface{}{
				"type": "text",
			},
		},
	}

	properties["chain"] = map[string]interface{}{
		"type":  "keyword",
		"index": true,
	}

	properties["contractAddress"] = map[string]interface{}{
		"type":  "text",
		"index": true,
	}

	properties["metadata"] = map[string]interface{}{
		"type":    "object",
		"enabled": false,
		"dynamic": true,
	}

	return map[string]interface{}{
		"properties": properties,
	}
}
