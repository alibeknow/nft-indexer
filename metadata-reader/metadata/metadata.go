package metadata

import (
	"context"
	"time"
)

type Metadata struct {
	ID        string     `bson:"_id"`
	Metadata  string     `bson:"metadata"`
	Type      string     `bson:"type"`
	CreatedAt *time.Time `bson:"created_at,omitempty"`
	UpdatedAt *time.Time `bson:"updated_at,omitempty"`
}

type DB interface {
	InsertOrUpdate(ctx context.Context, m Metadata) error
}
