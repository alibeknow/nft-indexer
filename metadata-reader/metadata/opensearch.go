package metadata

import (
	"encoding/json"
	"fmt"
)

type MetadataDocument struct {
	Name            string                 `json:"name"`
	Description     string                 `json:"description"`
	ContractAddress string                 `json:"contractAddress"`
	Attributes      []map[string]string    `json:"attributes"`
	Chain           Chain                  `json:"chain"`
	Metadata        map[string]interface{} `json:"metadata"`
}

func MapToMetadataDocument(contractAddress string, chain Chain, rawMetadata string) (MetadataDocument, error) {
	var metadata map[string]interface{}
	var props struct {
		Attributes []map[string]interface{} `json:"attributes"`
		Properties map[string]interface{}   `json:"properties"`
	}

	err := json.Unmarshal([]byte(rawMetadata), &metadata)
	if err != nil {
		return MetadataDocument{}, err
	}

	_ = json.Unmarshal([]byte(rawMetadata), &props)

	attributes := []map[string]string{}
	for _, attribute := range props.Attributes {
		var key interface{}
		if value, found := attribute["trait_type"]; found {
			key = value
		}

		attributes = append(attributes, map[string]string{
			"key":   fmt.Sprint(key),
			"value": fmt.Sprint(attribute["value"]),
		})
	}

	for key, value := range props.Properties {
		b, err := json.Marshal(value)
		if err != nil {
			return MetadataDocument{}, err
		}

		attributes = append(attributes, map[string]string{
			"key":   key,
			"value": string(b),
		})
	}

	name, _ := metadata["name"].(string)
	description, _ := metadata["description"].(string)

	return MetadataDocument{
		Name:            name,
		Description:     description,
		ContractAddress: contractAddress,
		Attributes:      attributes,
		Chain:           chain,
		Metadata:        metadata,
	}, nil
}
