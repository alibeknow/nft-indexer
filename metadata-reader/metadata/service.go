package metadata

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"go.uber.org/zap"

	"metadata-reader/private/httpreader"
	"metadata-reader/private/opensearch"
	"metadata-reader/queue"
)

var (
	metadataCounter = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "metadata_counter",
			Help: "Number of received metadata",
		},
	)
	errCounter = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "err_counter",
			Help: "Received errors count",
		},
	)
	tooManyReqCounter = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "too_many_request_counter",
			Help: "Number of rate limit requests",
		},
	)
	failedReqCounter = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "failed_request_counter",
			Help: "Number of failed requests",
		},
	)
	requestCounter = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "request_counter",
			Help: "Number of requests",
		},
	)
	storeMetadataDurationGauge = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "store_metadata_gauge",
			Help: "Store metadata duration",
		},
	)
)

func init() {
	prometheus.MustRegister(metadataCounter)
	prometheus.MustRegister(errCounter)
	prometheus.MustRegister(tooManyReqCounter)
	prometheus.MustRegister(failedReqCounter)
	prometheus.MustRegister(requestCounter)
	prometheus.MustRegister(storeMetadataDurationGauge)
}

type Config struct {
	Proxy   string
	Gateway string
	Retries int
	Timeout int
	Index   string
}

type Service struct {
	log        *zap.Logger
	db         DB
	httpReader *httpreader.HttpReader
	gateway    string
	opensearch *opensearch.OpenSearch
	indexName  string
}

func NewService(log *zap.Logger, config Config, db DB, search *opensearch.OpenSearch) (*Service, error) {
	var proxy *url.URL
	var err error

	if config.Proxy != "" {
		proxy, err = url.Parse(config.Proxy)
		if err != nil {
			return nil, err
		}
	}

	httpReader := httpreader.NewHttpReader(proxy, config.Retries, time.Duration(config.Timeout)*time.Second)

	return &Service{
		log:        log,
		db:         db,
		httpReader: httpReader,
		gateway:    config.Gateway,
		opensearch: search,
		indexName:  config.Index,
	}, nil
}

func (s *Service) Read(ctx context.Context, msg queue.MetadataMessage) (bool, error) {
	protocol := strings.SplitN(msg.TokenURI, ":", 2)[0]
	tokenRef := fmt.Sprintf("%s:%s:%s", msg.Blockchain, msg.ContractAddress, msg.TokenID)

	if protocol == "data" {
		tokenData := strings.SplitN(msg.TokenURI, ",", 2)
		if len(tokenData) < 2 {
			s.log.Debug("Inappropriate on-chain metadata format", zap.String("token", tokenRef))
			return false, nil
		}

		headWithoutDataPrefix := strings.Replace(tokenData[0], "data:", "", 1)

		dataTypeAndEncoding := strings.SplitN(headWithoutDataPrefix, ";", 2)
		if len(dataTypeAndEncoding) < 2 {
			s.log.Debug("Inappropriate on-chain metadata format", zap.String("token", tokenRef))
			return false, nil
		}

		if dataTypeAndEncoding[1] != "base64" {
			s.log.Debug("Inappropriate on-chain metadata encoding", zap.String("token", tokenRef))
			return false, nil
		}

		data, err := base64.StdEncoding.DecodeString(tokenData[1])
		if err != nil {
			s.log.Debug("Error decoding metadata", zap.String("token", tokenRef), zap.Error(err))
			return false, err
		}
		metadataCounter.Inc()
		startStoreToDB := time.Now()
		err = s.db.InsertOrUpdate(ctx, Metadata{
			ID:       tokenRef,
			Metadata: string(data),
			Type:     protocol,
		})
		storeMetadataDuration := time.Since(startStoreToDB).Microseconds()
		storeMetadataDurationGauge.Set(float64(storeMetadataDuration))
		if err != nil {
			s.log.Error("Error writing metadata to DB", zap.String("token", tokenRef), zap.Error(err))
			return false, err
		}

		if s.opensearch == nil {
			return false, nil
		}

		metadataDocument, err := MapToMetadataDocument(msg.ContractAddress, Chain(msg.Blockchain), string(data))
		if err != nil {
			s.log.Error("Error mapping metadata to opensearch document", zap.String("token", tokenRef), zap.Error(err))
			return false, err
		}

		b, err := json.Marshal(metadataDocument)
		if err != nil {
			s.log.Error("Error marshaling opensearch metadata document", zap.String("token", tokenRef), zap.Error(err))
			return false, err
		}

		_, err = s.opensearch.UpsertToIndex(ctx, s.indexName, tokenRef, strings.NewReader(string(b)))
		if err != nil {
			s.log.Error("Error writing metadata to opensearch index", zap.String("token", tokenRef), zap.Error(err))
			return false, err
		}

		s.log.Debug("Token is added to index", zap.String("token", tokenRef))

		return false, nil
	}

	uri := msg.TokenURI

	switch protocol {
	case "ipfs":
		trimmed := strings.TrimPrefix(uri, "ipfs://")
		trimmed = strings.TrimPrefix(trimmed, "ipfs/")
		uri = s.gateway + trimmed
	case "https":
		if strings.HasPrefix(msg.TokenURI, "https://ipfs.io/ipfs/") {
			trimmed := strings.TrimPrefix(msg.TokenURI, "https://ipfs.io/ipfs/")
			uri = s.gateway + trimmed
			break
		}
		if strings.HasPrefix(msg.TokenURI, "https://gateway.pinata.cloud/ipfs/") {
			trimmed := strings.TrimPrefix(msg.TokenURI, "https://gateway.pinata.cloud/ipfs/")
			uri = s.gateway + trimmed
			break
		}
		if strings.HasPrefix(msg.TokenURI, "https://api.opensea.io/api/") {
			replacer := strings.NewReplacer("{id}", strings.Replace(msg.TokenID, "0x", "", -1))
			uri = fmt.Sprintf(replacer.Replace(msg.TokenURI))
		}
	}

	resp, body, err := s.httpReader.Read(ctx, uri)
	requestCounter.Inc()
	if err != nil {
		s.log.Debug("HttpReader error", zap.String("token", tokenRef), zap.String("uri", uri), zap.String("originalUri", msg.TokenURI), zap.Error(err))
		errCounter.Inc()
		return false, err
	}

	if resp.StatusCode != http.StatusOK {
		s.log.Debug("Request failed with status code", zap.Int("statusCode", resp.StatusCode), zap.String("token", tokenRef), zap.String("uri", uri), zap.String("originalUri", msg.TokenURI))
		failedReqCounter.Inc()
		if resp.StatusCode == http.StatusTooManyRequests {
			tooManyReqCounter.Inc()
			return true, nil
		}
		return false, nil
	}

	metadataCounter.Inc()
	startStoreToDB := time.Now()
	err = s.db.InsertOrUpdate(ctx, Metadata{
		ID:       tokenRef,
		Metadata: string(body),
		Type:     protocol,
	})
	storeMetadataDuration := time.Since(startStoreToDB).Microseconds()
	storeMetadataDurationGauge.Set(float64(storeMetadataDuration))
	if err != nil {
		s.log.Error("Error writing metadata to DB", zap.String("token", tokenRef), zap.Error(err))
		return false, err
	}

	if s.opensearch == nil {
		return false, nil
	}

	metadataDocument, err := MapToMetadataDocument(msg.ContractAddress, Chain(msg.Blockchain), string(body))
	if err != nil {
		s.log.Error("Error mapping metadata to opensearch document", zap.String("token", tokenRef), zap.Error(err))
		return false, err
	}

	b, err := json.Marshal(metadataDocument)
	if err != nil {
		s.log.Error("Error marshaling opensearch metadata document", zap.String("token", tokenRef), zap.Error(err))
		return false, err
	}

	_, err = s.opensearch.UpsertToIndex(ctx, s.indexName, tokenRef, strings.NewReader(string(b)))
	if err != nil {
		s.log.Error("Error writing metadata to opensearch index", zap.String("token", tokenRef), zap.Error(err))
		return false, err
	}

	s.log.Debug("Token is added to index", zap.String("token", tokenRef))

	return false, nil
}
