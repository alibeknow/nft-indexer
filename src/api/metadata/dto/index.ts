export * from './nft-metadata-param.dto';
export * from './nfts-metadata-param.dto';
export * from './nfts-metadata-query.dto';
