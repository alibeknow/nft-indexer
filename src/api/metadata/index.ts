export * from './metadata.controller';
export * from './metadata.module';
export * from './metadata.service';
