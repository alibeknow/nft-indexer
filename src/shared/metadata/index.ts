export * from './metadata-repo-db';
export * from './metadata-repo-storage';
export * from './on-chain-reader';
export * from './ipfs-reader';
export * from './http-reader';
export * from './interfaces';
export * from './constants';
export * from './read-content-headers';
export * from './pinata-reader';
export * from './metadata-search';
