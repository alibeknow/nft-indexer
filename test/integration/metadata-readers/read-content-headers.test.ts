import {
  firstHeadResponseFixture,
  fourthHeadResponseFixture,
  secondHeadResponseFixture,
  thirdHeadResponseFixture,
} from '../fixtures/head-response.fixture';
import { ContentHeaders, readContentHeaders } from '@shared/metadata';
import nock, { ReplyHeaders } from 'nock';

jest.setTimeout(100000);

const tokenUrisMap = new Map<string, ReplyHeaders>([
  [ 'https://ipfs.io/ipfs/QmRdbJbYsB3CXufZ4CoPRHMFE8nJ8NXPPmijD5BVzT86qL', firstHeadResponseFixture ],
  [ 'https://tag-metadata.herokuapp.com/mutants/832', secondHeadResponseFixture ],
  [ 'https://assets.weedgang.game/genesis-strains/metadata/3615', thirdHeadResponseFixture ],
]);

const tokenUriWithoutContentLength = 'https://azimuth.network/erc721/1953660440.json';

describe('[read-content-headers]', () => {
  it('should get metadata content headers by tokenUri via head request', async () => {
    const result = new Map<string, ContentHeaders>();

    for (const [ tokenUri, headResponse ] of tokenUrisMap) {
      nock(tokenUri).head('').reply(200, { statusText: 'OK' }, headResponse);
      const contentHeaders = await readContentHeaders(tokenUri);
      result.set(tokenUri, contentHeaders);
    }

    result.forEach((contentHeaders) => {
      expect(typeof contentHeaders.type === 'string').toBeTruthy();
      expect(typeof contentHeaders.length === 'number').toBeTruthy();
    });

    expect(result).toMatchSnapshot();
  });

  it('should get metadata content headers even if there is no "content-length" header', async () => {
    nock(tokenUriWithoutContentLength).head('').reply(200, { statusText: 'OK' }, fourthHeadResponseFixture);
    const result = await readContentHeaders(tokenUriWithoutContentLength);

    expect(typeof result.type === 'string').toBeTruthy();
    expect(result.length).toBeNull();

    expect(result).toMatchSnapshot();
  });

  afterEach(async () => {
    nock.cleanAll();
  });
});
