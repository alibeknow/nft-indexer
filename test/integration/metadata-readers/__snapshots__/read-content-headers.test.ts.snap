// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[read-content-headers] should get metadata content headers by tokenUri via head request 1`] = `
Map {
  "https://ipfs.io/ipfs/QmRdbJbYsB3CXufZ4CoPRHMFE8nJ8NXPPmijD5BVzT86qL" => {
    "length": 10749707,
    "type": "image/jpeg",
  },
  "https://tag-metadata.herokuapp.com/mutants/832" => {
    "length": 344,
    "type": "application/json; charset=utf-8",
  },
  "https://assets.weedgang.game/genesis-strains/metadata/3615" => {
    "length": 1103,
    "type": "binary/octet-stream",
  },
}
`;

exports[`[read-content-headers] should get metadata content headers even if there is no "content-length" header 1`] = `
{
  "length": null,
  "type": "application/json",
}
`;
